storage/operation/children.go

```diff
@@ -9,27 +10,119 @@ import (
-//     serves as a reminder that the CALLER is responsible to ensure that the DEDUPLICATION CHECK is done elsewhere
-//     ATOMICALLY with this write operation.
+//   - The caller must acquire the [storage.LockInsertBlock] and hold it until the database write has been committed.
 //
-// No errors are expected during normal operation.
-func UpsertBlockChildren(lctx lockctx.Proof, w storage.Writer, blockID flow.Identifier, childrenIDs flow.IdentifierList) error {
-	held := lctx.HoldsLock(storage.LockInsertBlock) || lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock)
-	if !held {
-		return fmt.Errorf("missing required lock: %s or %s", storage.LockInsertBlock, storage.LockInsertOrFinalizeClusterBlock)
+// Expected error returns during normal operations:
+//   - [storage.ErrAlreadyExists] if the blockID is already indexed as a child of the parent
+func IndexNewBlock(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockInsertBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockInsertBlock)
 	}
 
-	return UpsertByKey(w, MakePrefix(codeBlockChildren, blockID), childrenIDs)
+	return insertNewBlock(rw, blockID, parentID)
+}
+
+// IndexNewRootBlock indexes a new root block by creating an empty children index for it.
+// This function is used for root blocks that have no parent and should not be added to any parent's children list.
+//
+// CAUTION:
+//   - The caller must acquire the [storage.LockInsertBlock] and hold it until the database write has been committed.
+//
+// No error returns are expected during normal operation.
+func IndexNewRootBlock(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockInsertBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockInsertBlock)
+	}
+
+	return insertNewBlockWithNoChild(rw, blockID)
+}
+
+// IndexNewClusterBlock indexes a new cluster block and updates the parent-child relationship in the block children index.
+// This function creates an empty children index for the new cluster block and adds the new block to the parent's children list.
+//
+// CAUTION:
+//   - The caller must acquire the [storage.LockInsertOrFinalizeClusterBlock] and hold it until the database write has been committed.
+//
+// Expected error returns during normal operations:
+//   - [storage.ErrAlreadyExists] if the blockID is already indexed as a child of the parent
+func IndexNewClusterBlock(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockInsertOrFinalizeClusterBlock)
+	}
+
+	return insertNewBlock(rw, blockID, parentID)
+}
+
+func insertNewBlock(rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
```

Comment:
I removed the procedure functions, and moved to operation instead, so that we can include all checks in the operation functions. 

Note this function does not require lock, and is private, so won't be accessable by outside callers. They have to use the public functions where locks are checked.